### PR TITLE
[BE-08] add sign in/sign up

### PR DESCRIPTION
**Pull request by dansob14**
https://trello.com/c/FdGCQHKJ/7-be-08-add-signin-signup-h
Use devise
Create an endpoint for sign_in
Create an endpoint for sign_up

AC:
1. User can register himself
2. User can log in himself
3. Email has to have validation

### CODE REVIEW

File: db/migrate/20220716120909_devise_token_auth_create_users.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+      t.boolean  :allow_password_change, :default => false
+
+      ## Rememberable
+      t.datetime :remember_created_at
+
+      ## Confirmable
+      t.string   :confirmation_token
+      t.datetime :confirmed_at
+      t.datetime :confirmation_sent_at
+      t.string   :unconfirmed_email # Only if using reconfirmable
```
Comment (plokijro) [inline db/migrate/20220716120909_devise_token_auth_create_users.rb]: Do we need `#Recoverable` and `#Confirmable` sections?

File: db/migrate/20220716120909_devise_token_auth_create_users.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      ## Lockable
+      # t.integer  :failed_attempts, :default => 0, :null => false # Only if lock strategy is :failed_attempts
+      # t.string   :unlock_token # Only if unlock strategy is :email or :both
+      # t.datetime :locked_at
+
+      ## User Info
+      t.string :name
+      t.string :nickname
+      t.string :image
+      t.string :email
```
Comment (plokijro) [inline db/migrate/20220716120909_devise_token_auth_create_users.rb]: Only `email` is needed

File: config/application.rb [hunk 3]
```diff
@@ -23,7 +23,15 @@ module Cookbook
...
-
+    config.middleware.use Rack::Cors do
+      allow do
+        origins '*'
+        resource '*',
+          headers: :any,
+          expose: ['access-token', 'expiry', 'token-type', 'uid', 'client'],
+          methods: [:get, :post, :options, :delete, :put]
+      end
+    end
```
Comment (plokijro) [inline config/application.rb]: I think it should be in `cors.rb` file

File: config/routes.rb [hunk 4]
```diff
@@ -1,4 +1,5 @@
 Rails.application.routes.draw do
+  mount_devise_token_auth_for 'User', at: 'auth'
```
Comment (plokijro) [inline config/routes.rb]: Move it to `api/v1` namespace and maybe create some specs 🤔

File: spec/requests/sign_in_spec.rb [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    end
+    
+    it 'check auth headers' do
+      login_user
+      expect(response.headers['uid']).to be_present
+      expect(response.headers['client']).to be_present
+      expect(response.headers['access-token']).to be_present
+    end
+  end
+  context 'with invalid params' do
```
Comment (plokijro) [inline spec/requests/sign_in_spec.rb]:
```suggestion
    end
  end
  
  context 'with invalid params' do
```

File: spec/requests/sign_in_spec.rb [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    let(:login_user) do
+      post '/api/v1/auth/sign_in', params: { email: user.email, password: 'InvalidPassword' }
+    end
```
Comment (plokijro) [inline spec/requests/sign_in_spec.rb]: Mov params to another `let` and then override only params

**Review by grabomarcin — APPROVED** [review]
Good job!

**Review by plokijro — APPROVED** [review]

File: config/initializers/cors.rb [hunk 7]
```diff
@@ -5,12 +5,12 @@
 
 # Read more: https://github.com/cyu/rack-cors
 
-Rails.application.config.middleware.insert_before 0, Rack::Cors do
+Rails.application.config.middleware.use Rack::Cors do
   allow do
-    origins ENV.fetch('API_ALLOWED_ORIGINS', '*').to_s.split(',')
+    origins '*'
```
Comment (DanielOldak) [inline config/initializers/cors.rb:10]: I think it shouldn't happen 😄

**Pull request merged**